fppanalysis/time_delay_estimation.py

```diff
@@ -0,0 +1,153 @@
+        scale: Distribution scale parameter.
+    """
+    import numpy as np
+    from scipy.stats import uniform, norm
+    from scipy.signal import fftconvolve
+    from scipy.optimize import minimize
+
+    def get_pdf(params, tauR):
+        if distribution == uniform:
+            pdf = np.zeros(len(tauR))
+            shape = 1 / (1 + params[1] ** 2)
+            low = params[0] * (1 - shape)
+            high = params[0] * (1 + shape)
+            pdf[np.logical_and(tauR < high, tauR >= low)] = 1 / (high - low)
+            return pdf
+        if distribution == norm:
+            return distribution.pdf(tauR, loc=params[0], scale=params[1])
+        if distribution.numargs > 0:
+            pdf = distribution.pdf(tauR, params[1], loc=0, scale=params[0])
+            if distribution == gamma and params[1] < 1:
+                pdf[tauR == 0] = 0
+            return pdf
+        else:
+            return distribution.pdf(tauR, loc=0, scale=params[0])
+
+    def get_average(params):
+        if distribution == uniform:
+            return params[0]
+        if distribution == norm:
+            return params[0]
+        if distribution.numargs > 0:
+            return distribution.stats(params[1], scale=scale, loc=0, moments="m")
+        else:
+            return distribution.stats(scale=scale, loc=0, moments="m")
+
+    def gamma_error(params):
+        pdf = get_pdf(params, ccf_times)
+        res = fftconvolve(est_acf, pdf, "same") * dt
+        res /= max(res)
+        return np.sum((res - est_ccf) ** 2)
+
+    def get_params():
+        if distribution in [uniform, norm]:
+            return [1, 1]
+        return np.ones(1 + distribution.numargs)
+
+    parameters = get_params()
+
+    ccf_times, est_ccf = cf.corr_fun(x, y, dt=dt, biased=False)
+    _, est_acf = cf.corr_fun(x, x, dt=dt, biased=False)
+
+    # We cut the cross correlation in the middle half to avoid noise near the ends
+    est_ccf = est_ccf[np.abs(ccf_times) < max(ccf_times) / 2]
+    est_acf = est_acf[np.abs(ccf_times) < max(ccf_times) / 2]
+    ccf_times = ccf_times[np.abs(ccf_times) < max(ccf_times) / 2]
+
+    max_cross_corr = max(ccf_times[np.argmax(est_ccf)], min_cutoff)
+    parameters[0] = max_cross_corr
+    # Time domain does not need to be full signal. This seems to be a good compromise.
+    domain_cutoff = max_cross_corr * 100
+
+    domain = np.abs(ccf_times) < domain_cutoff
+    ccf_times = ccf_times[domain]
+    est_ccf = est_ccf[domain]
+    est_acf = est_acf[domain]
+    est_ccf = est_ccf / max(est_ccf)
+
+    minimization = minimize(
+        gamma_error, parameters, method="Nelder-Mead", options={"maxiter": 10000}
+    )
+    if not minimization.success:
+        print("Optimization failed!!!")
+
+    num_args = len(minimization.x)
+    arg = minimization.x[:-1] if num_args > 1 else None
+    scale = minimization.x[0]
+    avg = get_average(minimization.x)
+
+    if ax is not None:
```

Comment:
Again, defining the plotting functionality in a separate function would make it more readable in my view.

Reply:
Done.